File: src/data/tutorials/047-connected-components.js
/**
 * Tutorial content for #047 — Connected Components
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `Given an integer n (nodes labelled 0 to n-1) and a list of undirected edges, return the number of connected components in the graph. A connected component is a group of nodes where every node can reach every other node.`,
    example: `n=5, edges=[[0,1],[1,2],[3,4]]\n→ Component 1: nodes {0,1,2} (all connected)\n→ Component 2: nodes {3,4} (connected to each other)\n→ Node clusters: 2 groups\n✅ Answer: 2`,
    keyInsight: `Start a counter at n (every node is its own component). Union-Find merges components as edges are processed — each successful union reduces the count by 1. Alternatively, DFS/BFS marks all reachable nodes; each new unvisited start is a new component.`,
  },

  approaches: {
    DFS: {
      intuition: `Iterate through all nodes. When you find an unvisited node, start a DFS from it, marking all reachable nodes as visited. Each time you start a new DFS (a new unvisited node), increment the component count. By the end, the count equals the number of connected components.`,
      steps: [
        `Build an undirected adjacency list from the edge list.`,
        `Create a visited array of size n, all false.`,
        `Initialise count = 0.`,
        `For each node i from 0 to n-1: if not visited, increment count and run dfs(i).`,
        `dfs(node): mark node visited, then recurse on all unvisited neighbours.`,
        `Return count.`,
      ],
      example: `n=5, edges=[[0,1],[1,2],[3,4]]\nAdj: 0:[1], 1:[0,2], 2:[1], 3:[4], 4:[3]\n\nNode 0 unvisited → count=1, dfs(0):\n  Visit 0. Neighbour 1 unvisited → dfs(1):\n    Visit 1. Neighbour 0 visited, neighbour 2 unvisited → dfs(2):\n      Visit 2. Neighbour 1 visited. Done.\nNode 1 visited. Node 2 visited.\nNode 3 unvisited → count=2, dfs(3):\n  Visit 3. Neighbour 4 unvisited → dfs(4): Visit 4.\nNode 4 visited.\n✅ Answer: 2`,
      keyInsight: `O(V+E) time, O(V+E) space. The classic "flood fill" approach — simple, intuitive, and works for any graph traversal problem.`,
    },

    BFS: {
      intuition: `Same idea as DFS but use a queue. For each unvisited node, start BFS, marking all reachable nodes as visited and incrementing the component count. BFS processes nodes level by level (by distance), but for component counting the order doesn't matter — both DFS and BFS explore the full component.`,
      steps: [
        `Build undirected adjacency list. Create visited array, all false. Initialise count = 0.`,
        `For each node i from 0 to n-1: if not visited, increment count and enqueue i.`,
        `BFS: while queue not empty, dequeue a node. Mark it visited. Enqueue all unvisited neighbours.`,
        `Continue until the queue empties (full component explored).`,
        `Return count.`,
      ],
      example: `n=5, edges=[[0,1],[1,2],[3,4]]\nAdj: 0:[1], 1:[0,2], 2:[1], 3:[4], 4:[3]\n\nNode 0 unvisited → count=1. Queue=[0]\nDequeue 0: visited. Neighbour 1 unvisited → enqueue. Queue=[1]\nDequeue 1: visited. Neighbour 0 visited, 2 unvisited → enqueue. Queue=[2]\nDequeue 2: visited. Neighbour 1 visited. Queue=[]\nNode 1,2 visited.\nNode 3 unvisited → count=2. Queue=[3]\nDequeue 3: visited. Neighbour 4 → enqueue. Queue=[4]\nDequeue 4: visited. Neighbour 3 visited. Queue=[]\n✅ Answer: 2`,
      keyInsight: `O(V+E) time, O(V+E) space. Preferred over DFS when recursion depth might be an issue (very large graphs). Both achieve the same result.`,
    },

    'Union-Find': {
      intuition: `Start with each node as its own component (count = n). For each edge, if the two nodes are in different components, merge them and decrement count. After processing all edges, count holds the number of components. Union-Find with path compression makes each operation nearly O(1).`,
      steps: [
        `Initialise parent[i] = i for all i. Set count = n.`,
        `Define find(x): follow parent pointers to root, applying path compression.`,
        `Define union(x, y): find roots of x and y. If different, set one as parent of the other and decrement count.`,
        `For each edge [u, v]: call union(u, v).`,
        `Return count.`,
      ],
      example: `n=5, edges=[[0,1],[1,2],[3,4]]\nParents: [0,1,2,3,4], count=5\n\nEdge [0,1]: find(0)=0, find(1)=1 → different → union, count=4. Parents=[0,0,2,3,4]\nEdge [1,2]: find(1)=0, find(2)=2 → different → union, count=3. Parents=[0,0,0,3,4]\nEdge [3,4]: find(3)=3, find(4)=4 → different → union, count=2. Parents=[0,0,0,3,3]\n✅ Answer: 2`,
      keyInsight: `O(E × α(N)) ≈ O(E) time with path compression. O(N) space. The most efficient approach for dynamic connectivity problems — handles edges one at a time without rebuilding the graph.`,
    },
  },
}

File: public/cpp/047-connected-components.cpp
/*
 * LeetCode Problem #323: Number of Connected Components in an Undirected Graph
 * Difficulty: Medium
 * Link: https://leetcode.com/problems/number-of-connected-components-in-an-undirected-graph/
 *
 * Problem Statement:
 * Given n nodes labeled 0 to n-1 and a list of undirected edges, find the number
 * of connected components in the graph.
 *
 * Example 1:
 *   Input: n = 5, edges = [[0,1],[1,2],[3,4]]
 *   Output: 2
 *
 * Example 2:
 *   Input: n = 5, edges = [[0,1],[1,2],[2,3],[3,4]]
 *   Output: 1
 *
 * Constraints:
 *   - 1 <= n <= 2000
 *   - 1 <= edges.length <= 5000
 *   - edges[i].length == 2
 *   - 0 <= ai < bi < n
 *   - No repeated edges
 */

#include <vector>
#include <queue>
#include <iostream>
using namespace std;

// ==================== APPROACH 1: DFS ====================
// DFS from each unvisited node; each DFS call = one new component.
// Time: O(V+E)  Space: O(V+E)
class Solution1 {
    void dfs(int node, vector<vector<int>>& adj, vector<bool>& vis) {
        vis[node] = true;
        for (int nb : adj[node]) if (!vis[nb]) dfs(nb, adj, vis);
    }
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> adj(n);
        for (auto& e : edges) { adj[e[0]].push_back(e[1]); adj[e[1]].push_back(e[0]); }
        vector<bool> vis(n, false);
        int count = 0;
        for (int i = 0; i < n; i++) if (!vis[i]) { dfs(i, adj, vis); count++; }
        return count;
    }
};

// ==================== APPROACH 2: BFS ====================
// BFS from each unvisited node; count how many BFS starts needed.
// Time: O(V+E)  Space: O(V+E)
class Solution2 {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<vector<int>> adj(n);
        for (auto& e : edges) { adj[e[0]].push_back(e[1]); adj[e[1]].push_back(e[0]); }
        vector<bool> vis(n, false);
        int count = 0;
        for (int i = 0; i < n; i++) {
            if (!vis[i]) {
                count++;
                queue<int> q;
                q.push(i); vis[i] = true;
                while (!q.empty()) {
                    int node = q.front(); q.pop();
                    for (int nb : adj[node]) if (!vis[nb]) { vis[nb] = true; q.push(nb); }
                }
            }
        }
        return count;
    }
};

// ==================== APPROACH 3: Union-Find (Optimal) ====================
// Union all edges; count distinct roots = number of components.
// Time: O(E·α(V))  Space: O(V)
class Solution3 {
    vector<int> parent, rank_;
    int find(int x) { return parent[x] == x ? x : parent[x] = find(parent[x]); }
    bool unite(int a, int b) {
        a = find(a); b = find(b);
        if (a == b) return false;
        if (rank_[a] < rank_[b]) swap(a, b);
        parent[b] = a;
        if (rank_[a] == rank_[b]) rank_[a]++;
        return true;
    }
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        parent.resize(n); rank_.assign(n, 0);
        for (int i = 0; i < n; i++) parent[i] = i;
        int components = n;
        for (auto& e : edges) if (unite(e[0], e[1])) components--;
        return components;
    }
};

void runTests() {
    Solution3 sol;
    vector<vector<int>> e1 = {{0,1},{1,2},{3,4}};
    cout << "Test 1: " << sol.countComponents(5, e1) << " (expected 2)\n";
    vector<vector<int>> e2 = {{0,1},{1,2},{2,3},{3,4}};
    cout << "Test 2: " << sol.countComponents(5, e2) << " (expected 1)\n";
    vector<vector<int>> e3 = {};
    cout << "Test 3: " << sol.countComponents(3, e3) << " (expected 3)\n";
}

int main() { runTests(); return 0; }

// Made with Bob
